### PR TITLE
HPCC-28217 HPCC HorizontalPodAutoScale

### PR DESCRIPTION
**Pull request by rpastrana**
- Creates helm helper template to add HPA
- Updates schema to allow HPA configuration
- Add default HPA for ESP,  Roxie

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: helm/hpcc/values.yaml [hunk 1]
```diff
@@ -748,6 +758,50 @@ roxie:
   #directAccessPlanes: []  #add direct access planes that roxie will read from without copying the data to its default data plane
   #ldapUser: roxie_file_access    #add system username for accessing files
   #egress: engineEgress
+  hpa:
```
Comment (rpastrana) [inline helm/hpcc/values.yaml:761]: @jakesmith this PR isn't done yet, and I don't intend to include this set of sample HPA metrics, but it illustrates the concept. It would all us to provide sensible default HPAs for various workload resources. Take a look a provide any feedback you might have. thanks

File: helm/hpcc/templates/_helpers.tpl [hunk 2]
```diff
@@ -2200,3 +2200,94 @@ Fills "result" dictionary with "planeCategories" and "namedPlanes"
...
+      - type: {{ $policy.type }}
+        value: {{ $policy.value }}
+        periodSeconds: {{ $policy.periodSeconds }}
+    {{- end }}
+  {{- end }}
+{{- end -}}
+{{- end -}}
+
+{{/*
+A template to generate a HorizontalPodAutoscaler for a named  workload resource (such as a Deployment or StatefulSet)
```
Comment (jakesmith) [inline helm/hpcc/templates/_helpers.tpl:2225]: trivial: double space after 'named'

File: helm/hpcc/templates/_helpers.tpl [hunk 3]
```diff
@@ -2200,3 +2200,94 @@ Fills "result" dictionary with "planeCategories" and "namedPlanes"
...
 {{- $_ := set .result "namedPlanes" $namedPlanes -}}
 {{- end -}}
+
+{{/*
+A template to generate a HorizontalPodAutoscaler for a named  workload resource (such as a Deployment or StatefulSet)
+Pass in dict with workload resource name and kind (Deployment|ReplicaSet|StatefulSet|ReplicationController), and hpa section
+*/}}
+{{- define "hpcc.addHPABehavior" }}
+{{- if .behavior -}}
+  {{- if hasKey .behavior "stabilizationWindowSeconds" }}
```
Comment (jakesmith) [inline helm/hpcc/templates/_helpers.tpl:2210]: minor: there may be exceptions already, but we have said we will use 1 space indentation for templating (vs yaml) in our helm chart templates.

File: helm/hpcc/templates/dafilesrv.yaml [hunk 4]
```diff
@@ -91,6 +91,10 @@ spec:
 {{ include "hpcc.addCertificateVolume" (dict "root" $ "name" .name "component" "dafilesrv" "certificate" .certificate "visibility" .service.visibility) | indent 6 }}
 {{- end }}
 ---
+{{- if hasKey . "hpa" -}}
+{{ include "hpcc.addHorizontalPodAutoscaler" (dict "name" .name "kind" "Deployment" "hpa" .hpa) }}
```
Comment (jakesmith) [inline helm/hpcc/templates/dafilesrv.yaml:95]: minor these and others could be indented, i.e.:
```
{{- if hasKey . "hpa" -}}
 {{- include "hpcc.addHorizontalPodAutoscaler" (dict "name" .name "kind" "Deployment" "hpa" .hpa) }}
{{- end }}

File: helm/hpcc/values.schema.json [hunk 5]
```diff
@@ -2946,6 +2967,131 @@
...
+            },
+            "scaleDown": {
+              "type": "object",
+              "$ref": "#/definitions/hpaScalingRules"
+            }
+          }
+        },
+        "metrics": {
+          "type": "array",
+          "description": "Specifies variables used to calculate the desired replica count. The desired replica count is calculated multiplying the ratio between the target value and the current value by the current number of pods. ",
```
Comment (jakesmith) [inline helm/hpcc/values.schema.json:2999]: trivial: extra space at end after 'pods.'

File: helm/hpcc/values.schema.json [hunk 6]
```diff
@@ -2946,6 +2967,131 @@
...
+                "type": "string",
+                "description": "Type is used to specify the scaling policy"
+              },
+              "value": {
+                "type": "integer",
+                "description": "Value contains the amount of change which is permitted by the policy. It must be greater than zero"
+              },
+              "periodSeconds": {
+                "type": "integer",
+                "description": "PeriodSeconds specifies the window of time for which the policy should hold true. PeriodSeconds must be greater than zero and less than or equal to 1800 (30 min)."
```
Comment (jakesmith) [inline helm/hpcc/values.schema.json:3023]: minor: casing. "Value" and "Type" also, vs 'selectPolicy' or 'selector' in descriptions below.

I suspect if going to mention the option in the description the casing should match, but it is probably best to describe without what it is without re-mentioning it.

File: helm/hpcc/values.schema.json [hunk 7]
```diff
@@ -2946,6 +2967,131 @@
...
+        },
+        "stabilizationWindowSeconds": {
+          "type": "integer",
+          "description": "Seconds for which past recommendations should be considered while scaling up or scaling down. StabilizationWindowSeconds must be greater than or equal to zero and less than or equal to 3600 (one hour). If not set, use the default values: - For scale up: 0 (i.e. no stabilization is done). - For scale down: 300 (i.e. the stabilization window is 300 seconds long)."
+        }
+      }
+    },
+    "hpaMetric" : {
+      "type": "object",
+      "description": "Target metric fo Horizontal Pod Autoscaler",
```
Comment (jakesmith) [inline helm/hpcc/values.schema.json:3040]: typo: "fo" -> "for"

File: helm/hpcc/values.schema.json [hunk 8]
```diff
@@ -2946,6 +2967,131 @@
...
+          "description": "selector is the string-encoded form of a standard kubernetes label selector for the given metric When set, it is passed as an additional parameter to the metrics server for more specific metrics scoping. When unset, just the metricName will be used to gather metrics.",
+          "properties": {
+            "matchLabels": {
+              "type": "object"
+             }
+          }
+        },
+        "describedObject": {
+          "type": "object",
+          "description": "specifies the descriptions of a object, such as kind,name apiVersion",
```
Comment (jakesmith) [inline helm/hpcc/values.schema.json:3062]: trivial: of an object. 

should that be "such as kind, name, apiVersion" ?

**Review by jakesmith — COMMENTED** [review]
@rpastrana - looks good overall, I think all or mostly minor/trivial comments

File: helm/examples/scaling/roxie-hpa-various-metrics.yaml [hunk 9]
```diff
@@ -0,0 +1,49 @@
+roxie:
```
[the remark this answers is not in the document]
Comment (rpastrana) [reply helm/examples/scaling/roxie-hpa-various-metrics.yaml:1]: not sure how that happened, btw, renaming to reflect it's only a sample of an HPA w/ multiple metrics, not an actual sample roxie hpa

File: helm/hpcc/templates/_helpers.tpl [hunk 10]
```diff
@@ -2200,3 +2200,94 @@ Fills "result" dictionary with "planeCategories" and "namedPlanes"
...
+      stabilizationWindowSeconds: {{ .behavior.stabilizationWindowSeconds }}
+  {{- end }}
+  {{- if hasKey .behavior "policies" }}
```
[the remark this answers is not in the document]
Comment (rpastrana) [reply helm/hpcc/templates/_helpers.tpl:2213]: the test is right, the name is misleading. Changing "behavior" to behaviorScale to represent the behavior.scaleUp/Down structure

**Comment by rpastrana** [comment]
@jakesmith please review, accidentally amended commit.

File: helm/hpcc/templates/_helpers.tpl [hunk 11]
```diff
@@ -2200,3 +2200,96 @@ Fills "result" dictionary with "planeCategories" and "namedPlanes"
...
+    apiVersion: apps/v1
+    kind: {{ .kind }}
+    name: {{ printf "%s" .name }}
+  minReplicas: {{ (hasKey .hpa "minReplicas") | ternary .hpa.minReplicas 1 }}
+  maxReplicas: {{ .hpa.maxReplicas }}
+   {{- if hasKey .hpa "behavior" }}
+  behavior: 
+    {{- if hasKey .hpa.behavior "scaleDown" }}
+    scaleDown:
+      {{ include "hpcc.addHPABehaviorScale" (dict "behaviorScale" .hpa.behavior.scaleDown) }}
```
Comment (jakesmith) [inline helm/hpcc/templates/_helpers.tpl:2246]: didn't notice on last review, but to make function portable (so they can be generated at any indentation level),
we make them generate without indentation, and then apply the correct level of indentation where they are included,
e.g. in this case (with indentation removed from hpcc.addHPABehaviorScale, this line becomes:
```
    {{- if hasKey .hpa.behavior "scaleDown" }}
    scaleDown:
     {{- include "hpcc.addHPABehaviorScale" (dict "behaviorScale" .hpa.behavior.scaleDown) | indent 6 }}
    {{- end }}
```

File: helm/examples/scaling/eclwatch-hpa-behavior.yaml [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    maxReplicas: 3
+    metrics:
+    - type: Resource
+      name: "cpu"
+      target: 
+        type: Utilization
+        value: "50"
+##Configures the scaling behavior of the target in both Up and Down directions
+    behavior: 
+      scaleDown:
```
Comment (jakesmith) [inline helm/examples/scaling/eclwatch-hpa-behavior.yaml:18]: shouldn't this also have a scaleUp to better illustrate it as example?

File: helm/hpcc/templates/dafilesrv.yaml [hunk 13]
```diff
@@ -91,6 +91,10 @@ spec:
 {{ include "hpcc.addCertificateVolume" (dict "root" $ "name" .name "component" "dafilesrv" "certificate" .certificate "visibility" .service.visibility) | indent 6 }}
 {{- end }}
 ---
+{{- if hasKey . "hpa" -}}
+ {{ include "hpcc.addHorizontalPodAutoscaler" (dict "name" .name "kind" "Deployment" "hpa" .hpa) }}
```
Comment (jakesmith) [inline helm/hpcc/templates/dafilesrv.yaml:95]: minor: white spacing in helm templates are horrible. This doesn't generate the wrong result, but it looks like it will indent by a space (which would be wrong), I think it would look more correct (and equal) if:
```
{{- if hasKey . "hpa" }}
 {{- include "hpcc.addHorizontalPodAutoscaler" (dict "name" .name "kind" "Deployment" "hpa" .hpa) }}
{{- end }}
```

File: helm/hpcc/values.schema.json [hunk 14]
```diff
@@ -2946,6 +2967,131 @@
...
+          "description" : "The metric type to determine workload scaling",
+          "enum": [ "Resource","Pods", "External", "Object"]
+        },
+        "name": {
+          "description": "The metric name",
+          "type": "string"
+        },
+        "selector": {
+          "type": "object",
+          "description": "selector is the string-encoded form of a standard kubernetes label selector for the given metric When set, it is passed as an additional parameter to the metrics server for more specific metrics scoping. When unset, just the metricName will be used to gather metrics.",
```
Comment (jakesmith) [inline helm/hpcc/values.schema.json:3053]: re:  https://github.com/hpcc-systems/HPCC-Platform/pull/17132#discussion_r1164378310

not a big deal I think, but were you going to normalize the capitalization in the descriptions that refer to the properties, e.g. this one "selector" is referred to in the correct case, but others e.g. "Value" and "Type" (lines 3013, 3017) are not. IMO, it is prob. best if the description describes without referring to the property name to avoid this issue.

**Review by jakesmith — CHANGES_REQUESTED** [review]
@rpastrana - a couple of new comments.

**Comment by rpastrana** [comment]
@jakesmith please review latest commit.
I do not plan to make further changes barring functional issues, or egregious stylistic issues.

**Review by jakesmith — COMMENTED** [review]
@rpastrana - looks good please squash (and commit message needs changing).

**Review by jakesmith — APPROVED** [review]
@rpastrana - looks good.

@ghalliday - Hopefully, this will allow others to think about how to use HPA's in HPCC and extend with custom metrics etc., so I think it is a good initial starting point.

File: helm/hpcc/templates/esp.yaml [hunk 15]
```diff
@@ -208,4 +211,5 @@ kind: ConfigMap
  {{ include "hpcc.addClientCertificate" (dict "root" $ "client" $remoteClient.name "organization" $remoteClient.organization "instance" $instance "component" $application "visibility" $visibility "secretTemplate" $remoteClient.secretTemplate) }}
 {{- end }}
 {{- end }}
+
```
Comment (ghalliday) [inline helm/hpcc/templates/esp.yaml:214]: unrelated?

**Review by ghalliday — COMMENTED** [review]
@rpastrana looks good.  A few minor comments.

File: helm/hpcc/templates/_helpers.tpl [hunk 16]
```diff
@@ -2200,3 +2200,99 @@ Fills "result" dictionary with "planeCategories" and "namedPlanes"
...
+        kind: {{ $metric.describedObject.kind }}
+        name: {{ $metric.describedObject.name }}
+     {{- end }}
+     {{- else if eq $metric.type "External"}}
+      metric:
+        name: {{ $metric.name }}
+      {{- if hasKey $metric "selector" }}
+        selector:
+          matchLabels:
+            {{ toYaml $metric.selector.matchLabels }}
```
[the remark this answers is not in the document]
Comment (rpastrana) [reply helm/hpcc/templates/_helpers.tpl]: Current output indentation seems correct (see below) but will force an indent
```
# Source: hpcc/templates/roxie.yaml
---
apiVersion: autoscaling/v2
kind: HorizontalPodAutoscaler
metadata:
  name: roxie-hpa
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: roxie
  minReplicas: 3
  maxReplicas: 15
  metrics:
  - type: Resource
    resource:
      name: cpu
      target:
        type: Utilization
        averageUtilization: 50
  - type: Pods
    pods:
      metric:
        name: packets-per-second
      target:
        type: AverageValue
        averageValue: 1k
  - type: External
    external:
      metric:
        name: queue_messages_ready
        selector:
          matchLabels:
            queue: worker_tasks
      target:
        type: AverageValue
        averageValue: 30
  - type: Object
    object:
      metric:
        name: requests-per-second
      describedObject:
        apiVersion: api/v1
        kind: Pod
        name: certain-pod-of-service-B
      target:
        type: AverageValue
        averageValue: 30
---

File: helm/hpcc/templates/_helpers.tpl [hunk 17]
```diff
@@ -2245,12 +2245,10 @@ spec:
    {{- if hasKey .hpa "behavior" }}
   behavior: 
     {{- if hasKey .hpa.behavior "scaleDown" }}
-    scaleDown:
-     {{- include "hpcc.addHPABehaviorScale" (dict "behaviorScale" .hpa.behavior.scaleDown) | indent 6 -}}
+    scaleDown: {{- include "hpcc.addHPABehaviorScale" (dict "behaviorScale" .hpa.behavior.scaleDown) | indent 6 }}
```
Comment (ghalliday) [inline helm/hpcc/templates/_helpers.tpl:2248]: This looks like it should be using nindent rather than indent.  Same for line 2251.  Is it correct?

File: helm/hpcc/templates/_helpers.tpl [hunk 18]
```diff
@@ -2264,21 +2262,21 @@ spec:
...
-     {{- end }}
+      {{- end }}
      {{- else if eq $metric.type "External"}}
       metric:
         name: {{ $metric.name }}
       {{- if hasKey $metric "selector" }}
         selector:
           matchLabels:
-            {{ toYaml $metric.selector.matchLabels }}
+{{ toYaml $metric.selector.matchLabels | indent 12}}
```
Comment (ghalliday) [inline helm/hpcc/templates/_helpers.tpl:2279]: minor: Better style is to indent as before, have a leading - and use nindent.
```
            {{- toYaml $metric.selector.matchLabels | nindent 12 }}
```

**Review by ghalliday — COMMENTED** [review]
@rpastrana looks close.  A couple of comments.

File: helm/hpcc/templates/_helpers.tpl (hunk 18 above)
Comment (rpastrana) [reply helm/hpcc/templates/_helpers.tpl:2279]: these are expensive style changes which amount to no-ops... will make change as suggested anyhow

Comment (rpastrana) [reply helm/hpcc/templates/_helpers.tpl:2279]: Suggested change applied, change now adheres to source code standard, and results in appropriately indented helm values:

```# Source: hpcc/templates/roxie.yaml
---
apiVersion: autoscaling/v2
kind: HorizontalPodAutoscaler
metadata:
  name: roxie-hpa
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: roxie
  minReplicas: 3
  maxReplicas: 15
  metrics:
  - type: Resource
    resource:
      name: cpu
      target:
        type: Utilization
        averageUtilization: 50
  - type: Pods
    pods:
      metric:
        name: packets-per-second
      target:
        type: AverageValue
        averageValue: 1k
  - type: External
    external:
      metric:
        name: queue_messages_ready
        selector:
          matchLabels:
            queue: worker_tasks
      target:
        type: AverageValue
        averageValue: 30
  - type: Object
    object:
      metric:
        name: requests-per-second
      describedObject:
        apiVersion: api/v1
        kind: Pod
        name: certain-pod-of-service-B
      target:
        type: AverageValue
        averageValue: 30
---

File: helm/hpcc/templates/_helpers.tpl (hunk 17 above)
Comment (rpastrana) [reply helm/hpcc/templates/_helpers.tpl]: Yes it works as-is, I tested several times before PR created, several other times after your comment. See output below. Spent more time shifting the whitechar controls for generated output until the nindent based output resulted in the same format as my original layout.
```---
# Source: hpcc/templates/esp.yaml
apiVersion: autoscaling/v2
kind: HorizontalPodAutoscaler
metadata:
  name: eclwatch-hpa
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: eclwatch
  minReplicas: 1
  maxReplicas: 3
  behavior:
    scaleDown:
      stabilizationWindowSeconds: 300
      policies:
      - type: Pods
        value: 1
        periodSeconds: 20
      - type: Percent
        value: 10
        periodSeconds: 60
    scaleUp:
      stabilizationWindowSeconds: 0
      policies:
      - type: Percent
        value: 10
        periodSeconds: 60
      - type: Pods
        value: 3
        periodSeconds: 60
      selectPolicy: Min
  metrics:
  - type: Resource
    resource:
      name: cpu
      target:
        type: Utilization
        averageUtilization: 50
---

**Comment by rpastrana** [comment]
@ghalliday please review.

**Review by ghalliday — APPROVED** [review]
Thanks.  Please squash.

**Comment by rpastrana** [comment]
> Thanks. Please squash.

Done. Thanks.
